clang/lib/Serialization/ASTWriter.cpp

```diff
@@ -3896,7 +3903,7 @@ void ASTWriter::WriteIdentifierTable(Preprocessor &PP,
 
       // Write out identifiers if either the ID is local or the identifier has
       // changed since it was loaded.
-      if (ID >= FirstIdentID || !Chain || !II->isFromAST() ||
+      if (isLocalIdentifierID(ID) || !Chain || !II->isFromAST() ||
```

Comment:
This diff makes it seem that `ID >= FirstIdentID` is equivalent to `isLocalIdentifierID(ID)`, which I don't think is the case. Can you explain what's going on here?

Reply:
Maybe it is because the name `FirstIdentID` is confusing. The proper name may be `FirstLocalIdentID`.

Previously, before this patch, `FirstIdentID` will be `Number of identifier ID` + 1. This is the reason why the encoding of Identifier IDs may be affected by imported modules. This is also the major motivation of the patch. 

So all the ID bigger or equal to FirstIdentID must be the ID for local identifiers previously. But it won't be the case we make this patch.

Reply:
Ah, got it. I was looking at this line in `ASTWriter.h`:

```c++
  /// The first ID number we can use for our own identifiers.
  serialization::IdentID FirstIdentID = serialization::NUM_PREDEF_IDENT_IDS;
```

and didn't realize it's being re-initialized in `ASTWriter::ReaderInitialized()`.

Doesn't that make the `!II->isFromAST()` check redundant then?

Reply:
Nice catch! The  `!II->isFromAST()` check looks redundant indeed. I've made a NFC patch to address this:  https://github.com/llvm/llvm-project/commit/799ae77993fa5d7b0638f10b3895090f8748de92